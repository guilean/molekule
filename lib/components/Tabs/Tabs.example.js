import React from 'react';
import Tabs from './Tabs';
import LiveEdit from '../../live-edit';

const TabsCode = `<Tabs
  tabs={[
    {
      id: 'test-one',
      title: 'Tab One',
      content: <div>Hey, I'm the first panel</div>,
    },
    {
      id: 'test-two',
      title: 'Tab Two',
      content: <div>Hey, I'm the second panel</div>,
    }, {
      id: 'test-three',
      title: 'Tab Three',
      content: <div>Hey, I'm the third panel</div>,
      disabled: true
    }]}
/>`;

const TabsCodeVertical = `<Tabs
  vertical
  tabs={[
    {
      id: 'test-one',
      title: 'Tab One',
      content: <div>Hey, I'm the first panel</div>,
    },
    {
      id: 'test-two',
      title: 'Tab Two',
      content: <div>Hey, I'm the second panel</div>,
    }]}
/>`;

export default {
  group: 'Tabs',
  render: () => (
    <React.Fragment>
      <p>
        Tabs enable a user to quickly switch between several panels. They can be laid out either horizontally or
        vertically depending on the <code>vertical</code> prop.
      </p>

      <p>
        Tabs receives a <code>tabs</code> prop, which is an array of tab objects that look like the following:
      </p>
      <pre>
        <code>
          {JSON.stringify(
            {
              id: 'string',
              title: 'string|HTMLNode|ReactElement',
              content: 'string|HTMLNode|ReactElement',
              disabled: 'boolean',
              onActive: 'function',
            },
            null,
            2
          )}
        </code>
      </pre>

      <h2>Layout</h2>
      <h3>Horizontal</h3>
      <LiveEdit code={TabsCode} scope={{ Tabs }} />

      <h3>Vertical</h3>
      <LiveEdit code={TabsCodeVertical} scope={{ Tabs }} />

      <h2>Callbacks</h2>
      <h3>onChange</h3>
      <p>
        Tabs, by default, are self-contained and will control themselves. However, if an <code>onChange</code> prop is
        provided, Tabs can now be controlled by the <code>active</code>
      </p>

      <h3>onActive</h3>
      <p>
        <code>onActive</code> will be fired when a tab becomes active. Usually this is used to activate some async
        action to load data or for analytics;
      </p>
    </React.Fragment>
  ),
};
