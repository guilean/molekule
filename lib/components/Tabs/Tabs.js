import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const TabsProvider = styled.div`
  display: flex;
  flex-direction: ${p => (p.vertical ? 'row' : 'column')};
`;

const TabList = styled.ul`
  display: flex;
  margin: 0;
  padding: 0;
  list-style: none;
  flex-direction: ${p => (p.vertical ? 'column' : 'row')};
  border-${p => (p.vertical ? 'right' : 'bottom')}: 1px solid #dfe3e8;
`;

const TabListItem = styled.li`
  display: flex;
  margin: 0;
  padding: 0;
`;

const Tab = styled.button`
  appearance: none;
  margin: 0;
  padding: 0;
  background: none;
  border: none;
  font-size: 14px;
  position: relative;
  justify-content: center;
  margin-${p => (p.vertical ? 'left' : 'top')}: 1px;
  margin-${p => (p.vertical ? 'right' : 'bottom')}: -1px;
  padding: ${p => (p.vertical ? '0' : '0 12px')};
  outline: none;
  white-space: nowrap;
  cursor: pointer;
`;

const TabTitle = styled.span`
  color: ${p => (p.active ? '#212b36' : '#637381')};
  display: block;
  padding: ${p => (p.vertical ? '8px 12px' : '12px 8px')};
  border-${p => (p.vertical ? 'right' : 'bottom')}: 4px solid ${p =>
  (p.active ? p.theme.colors.primary : 'transparent')};

  &::hover {
    color: #212b36;
    border-color: ${p => (p.active ? p.theme.colors.primary : '#d9dee3')};
  }
`;

const TabContent = styled.div`
  padding: 12px;
`;

class Tabs extends Component {
  static propTypes = {
    tabs: PropTypes.arrayOf(PropTypes.shape()).isRequired,
    active: PropTypes.number,
    vertical: PropTypes.bool,
    onSelect: PropTypes.func,
  };

  static defaultProps = {
    active: undefined,
    vertical: false,
  };

  constructor(props) {
    super(props);

    this.state = {
      active: props.active || 0,
    };
  }

  componentDidUpdate() {
    if (this.props.active && this.props.active !== this.state.active) {
      this.setState({
        active: this.props.active,
      });
    }
  }

  get controlled() {
    return 'active' in this.props && 'onSelect' in this.props;
  }

  handleSelect = selected => {
    if (this.controlled) {
      this.props.onSelect(selected);
    } else {
      this.setState({
        active: selected,
      });
    }
  };

  renderHeadings() {
    return this.props.tabs.map((tab, i) => (
      <TabListItem key={tab.id || i}>
        <Tab vertical={this.props.vertical} onClick={() => this.handleSelect(i)}>
          <TabTitle active={this.state.active === i} vertical={this.props.vertical}>
            {tab.title}
          </TabTitle>
        </Tab>
      </TabListItem>
    ));
  }

  renderContent() {
    const { content } = this.props.tabs[this.state.active];
    if (React.isValidElement(content)) {
      return content;
    }
    throw new Error('Tab content must be a valid React element.');
  }

  render() {
    return (
      <TabsProvider vertical={this.props.vertical}>
        <TabList vertical={this.props.vertical}>{this.renderHeadings()}</TabList>
        <TabContent>{this.renderContent()}</TabContent>
      </TabsProvider>
    );
  }
}

export default Tabs;
