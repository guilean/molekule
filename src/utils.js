import get from 'lodash/get';
import kebabCase from 'lodash/kebabCase';
import styled from 'styled-components';

export const getFromTheme = (theme, lookup, fallback) => get(theme, lookup, fallback);

export const getComponentVariant = (theme, key, variant) => {
  const config = getFromTheme(theme, `components.${key}.variants.${variant}`, theme.variants[variant]);
  if (!config) throw new Error(`Refractal: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentStyle = key => ({ theme }) => getFromTheme(theme, `components.${key}.style`, {});

export const getComponentClassName = name => ({ theme: { classPrefix }, variant }) =>
  `${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

export const createComponent = ({ name, tag = 'div', as }) => {
  const component = as ? styled(as) : styled[tag];

  return component.attrs({
    className: getComponentClassName(kebabCase(name)),
  })`
    ${getComponentStyle(name)}
    ${({ styles = {} }) => styles[name] || {}}
  `;
};
