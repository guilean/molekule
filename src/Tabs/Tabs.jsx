import React, { Component } from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import { createComponent } from '../utils';

const TabsProvider = createComponent({
  name: 'Tabs',
}).extend`
  display: flex;
  flex-direction: ${p => (p.vertical ? 'row' : 'column')};
`;

const TabList = createComponent({
  name: 'TabList',
  tag: 'ul',
}).extend`
  display: flex;
  margin: 0;
  padding: 0;
  list-style: none;
  flex-direction: ${p => (p.vertical ? 'column' : 'row')};
  ${p =>
    !p.vertical &&
    css`
      border-bottom: 1px solid #dfe3e8;
    `};
`;

const TabListItem = createComponent({
  name: 'TabListItem',
  tag: 'li',
}).extend`
  display: flex;
  margin: 0;
  padding: 0;
`;

const Tab = createComponent({
  name: 'Tab',
  tag: 'button',
}).extend`
  appearance: none;
  margin: 0;
  padding: 0;
  background: none;
  border: none;
  font-size: 14px;
  font-family: inherit;
  position: relative;
  justify-content: center;
  cursor: pointer;
  margin-${p => (p.vertical ? 'left' : 'top')}: 1px;
  margin-${p => (p.vertical ? 'right' : 'bottom')}: -1px;
  padding: ${p => (p.vertical ? '0' : '0 12px')};
  outline: none;
  white-space: nowrap;
  opacity: ${p => (p.disabled ? 0.5 : 1)};

  &:active {
    outline: none;
  }
`;

const TabTitle = createComponent({
  name: 'TabTitle',
  tag: 'span',
}).extend`
  color: ${p => (p.active ? '#212b36' : '#637381')};
  display: block;
  padding: ${p => (p.vertical ? '8px 12px 8px 8px' : '12px 8px')};
  border-${p => (p.vertical ? 'left' : 'bottom')}: 4px solid ${p =>
  p.active ? p.theme.colors.primary : 'transparent'};

  &::hover {
    color: #212b36;
    border-color: ${p => (p.active ? p.theme.colors.primary : '#d9dee3')};
  }
`;

const TabContent = createComponent({
  name: 'TabContent',
  tag: 'div',
}).extend`
  width: 100%;
`;

class Tabs extends Component {
  static propTypes = {
    tabs: PropTypes.arrayOf(PropTypes.shape()).isRequired,
    active: PropTypes.number,
    vertical: PropTypes.bool,
    onChange: PropTypes.func,
  };

  static defaultProps = {
    vertical: false,
  };

  state = {
    active: this.props.active || 0,
  };

  componentDidUpdate() {
    if (this.props.active && this.props.active !== this.state.active) {
      this.setState({
        active: this.props.active,
      });
    }
  }

  handleTabClick = selected => {
    if (this.props.onChange) {
      this.props.onChange(selected);
    }

    this.setState({
      active: selected,
    });
  };

  renderHeadings() {
    return this.props.tabs.map((tab, i) => {
      const id = tab.id || i;

      return (
        <TabListItem key={id}>
          <Tab
            id={id}
            disabled={tab.disabled}
            vertical={this.props.vertical}
            onClick={() => {
              if (tab.onActive) {
                tab.onActive();
              }
              this.handleTabClick(i);
            }}>
            <TabTitle active={this.state.active === i} vertical={this.props.vertical}>
              {tab.title}
            </TabTitle>
          </Tab>
        </TabListItem>
      );
    });
  }

  renderContent() {
    const { content } = this.props.tabs[this.state.active];
    if (React.isValidElement(content)) {
      return content;
    }
    throw new Error('Tab content must be a valid React element.');
  }

  render() {
    return (
      <TabsProvider vertical={this.props.vertical}>
        <TabList vertical={this.props.vertical}>{this.renderHeadings()}</TabList>
        <TabContent>{this.renderContent()}</TabContent>
      </TabsProvider>
    );
  }
}

export default Tabs;
