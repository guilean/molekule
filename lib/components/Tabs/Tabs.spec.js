import React from 'react';
import { mountWithTheme } from '../../../test/utils';
import Tabs from './Tabs';

const TabContent = ({ body }) => <div>This is the content for Tab {body}</div>;

const TABS = [
  {
    id: 'test-one',
    title: 'Tab One',
    content: <TabContent body="one" />,
  },
  {
    id: 'test-two',
    title: 'Tab Two',
    content: <TabContent body="two" />,
  },
];

test('Card', () => {
  const component = mountWithTheme(<Tabs tabs={TABS} />);

  expect(component).toMatchSnapshot();
});
