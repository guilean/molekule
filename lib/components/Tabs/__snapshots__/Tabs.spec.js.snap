// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> renders 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  padding: 0;
  list-style: none;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  border-bottom: 1px solid #dfe3e8;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  padding: 0;
}

.c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  margin: 0;
  padding: 0;
  background: none;
  border: none;
  font-size: 14px;
  position: relative;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  margin-top: 1px;
  margin-bottom: -1px;
  padding: 0 12px;
  outline: none;
  white-space: nowrap;
  opacity: 1;
}

.c3:active {
  outline: none;
}

.c4 {
  color: #212b36;
  display: block;
  padding: 12px 8px;
  border-bottom: 4px solid #5759D6;
}

.c4::hover {
  color: #212b36;
  border-color: #5759D6;
}

.c5 {
  color: #637381;
  display: block;
  padding: 12px 8px;
  border-bottom: 4px solid transparent;
}

.c5::hover {
  color: #212b36;
  border-color: #d9dee3;
}

.c6 {
  padding: 12px;
}

<Tabs
  tabs={
    Array [
      Object {
        "content": <TabContent
          body="one"
        />,
        "id": "test-one",
        "title": "Tab One",
      },
      Object {
        "content": <TabContent
          body="two"
        />,
        "id": "test-two",
        "onActive": [MockFunction],
        "title": "Tab Two",
      },
    ]
  }
  vertical={false}
>
  <styled.div
    vertical={false}
  >
    <div
      className="c0"
    >
      <styled.ul
        vertical={false}
      >
        <ul
          className="c1"
        >
          <styled.li
            key="test-one"
          >
            <li
              className="c2"
            >
              <styled.button
                id="test-one"
                onClick={[Function]}
                vertical={false}
              >
                <button
                  className="c3"
                  id="test-one"
                  onClick={[Function]}
                >
                  <styled.span
                    active={true}
                    vertical={false}
                  >
                    <span
                      className="c4"
                    >
                      Tab One
                    </span>
                  </styled.span>
                </button>
              </styled.button>
            </li>
          </styled.li>
          <styled.li
            key="test-two"
          >
            <li
              className="c2"
            >
              <styled.button
                id="test-two"
                onClick={[Function]}
                vertical={false}
              >
                <button
                  className="c3"
                  id="test-two"
                  onClick={[Function]}
                >
                  <styled.span
                    active={false}
                    vertical={false}
                  >
                    <span
                      className="c5"
                    >
                      Tab Two
                    </span>
                  </styled.span>
                </button>
              </styled.button>
            </li>
          </styled.li>
        </ul>
      </styled.ul>
      <styled.div>
        <div
          className="c6"
        >
          <TabContent
            body="one"
          >
            <div>
              This is the content for Tab 
              one
            </div>
          </TabContent>
        </div>
      </styled.div>
    </div>
  </styled.div>
</Tabs>
`;
